store/src/java/com/zimbra/cs/service/mail/SearchAction.java

```diff
@@ -104,4 +112,84 @@ private static void performMoveAction(BulkAction action, SearchRequest searchReq
             }
         }
     }
+
+    private static void performFlagAction(BulkAction action, SearchRequest searchRequest,
+        List<SearchHit> searchHits, Mailbox mbox, OperationContext octxt) throws ServiceException {
+        boolean markRead;
+        if (action.getMarkRead() != null) {
+            markRead = action.getMarkRead();
+        } else {
+            throw ServiceException.INVALID_REQUEST("read value not provided", null);
+        }
+        for (SearchHit searchHit : searchHits) {
+            String id = searchHit.getId();
+            if ("message".equalsIgnoreCase(searchRequest.getSearchTypes())) {
+                Message msg = mbox.getMessageById(octxt, Integer.parseInt(id));
+                if (markRead && msg.isUnread() && !RedoLogProvider.getInstance().isSlave()) {
+                    try {
```

Comment:
These 4 try/catch blocks are very similar.  Would be better to re-use code by creating a method with parameters for the differences.

Reply:
Done